cfgmgr/natmgr.cpp

```diff
@@ -0,0 +1,6391 @@
+{
+    /* Get all Static NAT entries */
+    for (auto it = m_staticNatEntry.begin(); it != m_staticNatEntry.end(); it++)
+    {
+        /* Ensure that interface is having some port, otherwise Entry is not yet added */
+        if (((*it).first == global_ip) && ((*it).second.local_ip == local_ip) &&
+            ((*it).second.interface != NONE_STRING))
+        {
+            /* Matches with the Static NAT entry */
+            return true;
+        }
+    }
+    return false;
+}
+
+/* To get the configured interface for the given global ip address */
+bool NatMgr::getIpEnabledIntf(const string &global_ip, string &interface)
+{
+    /* Get all Ports from Ip Interface Info */
+    for (auto it = m_natIpInterfaceInfo.begin(); it != m_natIpInterfaceInfo.end(); it++)
+    {
+        /* Get all IpAddress from Ip Interface Values from the key (Port) */
+        for (auto ipPrefix = m_natIpInterfaceInfo[(*it).first].begin(); ipPrefix != m_natIpInterfaceInfo[(*it).first].end(); ipPrefix++)
+        {
+            /* Check the global ip address is in subnet */
+            if (isGlobalIpMatching(*ipPrefix, global_ip) == true)
+            {
+                /* Matched with this interface, return now */
+                interface = (*it).first;
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+/* This is ideally called on docker stop  */
+void NatMgr::cleanupPoolIpTable(void)
+{
+    SWSS_LOG_INFO("Cleaning the NAPT Pool IP table from APP_DB");
+    for (auto it = m_natPoolInfo.begin(); it != m_natPoolInfo.end(); it++)
+    {
+        /* Delete pool ip from APPL_DB */
+        setNaptPoolIpTable(DELETE, ((*it).second).ip_range, ((*it).second).port_range);
+    }
+}
+
+/* To Add/Delete NAPT pool ip table to APPL_DB */
+void NatMgr::setNaptPoolIpTable(const string &opCmd, const string &ip_range, const string &port_range)
+{
+    uint32_t ipv4_addr_low, ipv4_addr_high, ip, setIp;
+    char ipAddr[INET_ADDRSTRLEN];
+    std::vector<swss::FieldValueTuple> values;
+
+    if (!port_range.empty() and (port_range != "NULL"))
+    {
+        swss::FieldValueTuple p("port_range", port_range);
+        values.push_back(p);
+        vector<string> nat_ip = tokenize(ip_range, range_specifier);
+
+        /* Check the pool is valid */
+        if (nat_ip.empty())
+        {
+            SWSS_LOG_INFO("NAT pool is not valid");
+            return;
+        }
+        else if (nat_ip.size() == 2)
+        {
+            inet_pton(AF_INET, nat_ip[1].c_str(), &ipv4_addr_high);
+            ipv4_addr_high = ntohl(ipv4_addr_high);
+            inet_pton(AF_INET, nat_ip[0].c_str(), &ipv4_addr_low);
+            ipv4_addr_low = ntohl(ipv4_addr_low);
+        }
+        else
+        {
+            inet_pton(AF_INET, nat_ip[0].c_str(), &ipv4_addr_low);
+            ipv4_addr_high = ntohl(ipv4_addr_low);
+            ipv4_addr_low = ntohl(ipv4_addr_low);
+        }
+
+        for (ip = ipv4_addr_low; ip <= ipv4_addr_high; ip++)
```

Comment:
Why set like this and not as a range?

Reply:
Here m_appNaptPoolIpTable is used by NatSyncd code to know whether the packet notified by conntrack table is NAPT'ed or NAT'ed.

For example, Create a NAT pool to do NAPT'ed like "config nat add pool p1 65.55.42.5 100-200"
If packet comes to cpu with Src Ip as 12.0.0.1 and Src port as 100, it get's SNAPT'ed to Src Ip as 65.55.42.5 and Src Port as 100, kernel picks the "100" as it is available from Pool.
So kernel doesn't set any separate flags in the conntrack entry if it is NAPT'ed. Only SNAT/DNAT flags are set in the conntrack entry that is notified to natsyncd.
When the translated port number is the same, Natsyncd will not be able to say if it is NAT'ed or NAPT'ed. So the NaptPoolIpTable is added for every IP address in the pool to help NatSyncd pick the conntrack notifications as NAPT and add them as NAPT entries in the APP_DB.